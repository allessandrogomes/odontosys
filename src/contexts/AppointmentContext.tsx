import { createContext, useReducer, useContext, ReactNode } from "react"

type Appointment = {
    patientName: string | null
    patientId: number | null
    dentistName: string | null
    dentistId: number | null
    scheduledAt: string | null
    endsAt: string | null
    durationMinutes: number | null
    procedure: string | null
    step: number
}

type State = Appointment

type Action =
    | { type: "SET_PATIENT"; payload: { id: number | null; name: string | null } }
    | { type: "SET_DENTIST"; payload: { id: number | null; name: string | null } }
    | { type: "SET_SCHEDULE"; payload: { scheduledAt: string | null; endsAt: string | null } }
    | { type: "SET_PROCEDURE"; payload: { procedure: string | null; durationMinutes: number | null }  }
    | { type: "SET_STEP"; payload: number }
    | { type: "RESET_APPOINTMENT" }

const initialState: State = {
    patientName: null,
    patientId: null,
    dentistName: null,
    dentistId: null,
    scheduledAt: null,
    endsAt: null,
    durationMinutes: null,
    procedure: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_PATIENT":
            return {
                ...state,
                patientId: action.payload.id,
                patientName: action.payload.name,
            }
        case "SET_DENTIST":
            return {
                ...state,
                dentistId: action.payload.id,
                dentistName: action.payload.name,
            }
        case "SET_SCHEDULE":
            return {
                ...state,
                scheduledAt: action.payload.scheduledAt,
                endsAt: action.payload.endsAt,
            }
        case "SET_PROCEDURE":
            return {
                ...state,
                procedure: action.payload.procedure,
                durationMinutes: action.payload.durationMinutes,
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        case "RESET_APPOINTMENT":
            return initialState
        default:
            return state
    }
}

const AppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function AppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <AppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </AppointmentContext.Provider>
    )
}

export function useAppointmentContext() {
    const context = useContext(AppointmentContext)
    if (!context) {
        throw new Error("useAppointment must be used within an AppointmentProvider")
    }
    return context
}