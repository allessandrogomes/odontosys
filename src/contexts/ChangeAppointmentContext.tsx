import { createContext, useReducer, useContext, ReactNode } from "react"

type ChangeAppointmentState = {
    appointments: IAppointment[] | [] | null
    selectedAppointment: IAppointment | null
    step: number
}

type State = ChangeAppointmentState

type Action =
    | { type: "SET_APPOINTMENTS"; payload: IAppointment[] | [] | null }
    | { type: "SET_SELECTED_APPOINTMENT"; payload: IAppointment | null }
    | { type: "SET_STEP"; payload: number }

const initialState: State = {
    appointments: null,
    selectedAppointment: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_APPOINTMENTS":
            return {
                ...state,
                appointments: action.payload
            }
        case "SET_SELECTED_APPOINTMENT":
            return {
                ...state,
                selectedAppointment: action.payload
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        default:
            return state
    }
}

const ChangeAppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function ChangeAppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <ChangeAppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </ChangeAppointmentContext.Provider>
    )
}

export function useChangeAppointmentContext() {
    const context = useContext(ChangeAppointmentContext)
    if (!context) {
        throw new Error("useChangeAppointmentContext must be used within and ChangeAppointmentProvider")
    }
    return context
}