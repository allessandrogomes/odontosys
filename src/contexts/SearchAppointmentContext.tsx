import { createContext, useReducer, useContext, ReactNode } from "react"

type SearchAppointmentState = {
    appointments: IAppointment[] | [] | null
    selectedAppointment: IAppointment | null
    step: number
}

type State = SearchAppointmentState

type Action =
    | { type: "SET_APPOINTMENTS"; payload: IAppointment[] | [] | null }
    | { type: "SET_SELECTED_APPOINTMENT"; payload: IAppointment | null }
    | { type: "SET_STEP"; payload: number }

const initialState: State = {
    appointments: null,
    selectedAppointment: null,
    step: 1
}

function reducer(state: State, action: Action): State {
    switch (action.type) {
        case "SET_APPOINTMENTS":
            return {
                ...state,
                appointments: action.payload
            }
        case "SET_SELECTED_APPOINTMENT":
            return {
                ...state,
                selectedAppointment: action.payload
            }
        case "SET_STEP":
            return {
                ...state,
                step: action.payload
            }
        default:
            return state
    }
}

const SearchAppointmentContext = createContext<{
    state: State
    dispatch: React.Dispatch<Action>
} | undefined>(undefined)

export function SearchAppointmentProvider({ children }: { children: ReactNode }) {
    const [state, dispatch] = useReducer(reducer, initialState)

    return (
        <SearchAppointmentContext.Provider value={{ state, dispatch }}>
            {children}
        </SearchAppointmentContext.Provider>
    )
}

export function useSearchAppointmentContext() {
    const context = useContext(SearchAppointmentContext)
    if (!context) {
        throw new Error("useSearchAppointment must be used within an SearchAppointmentProvider")
    }
    return context
}