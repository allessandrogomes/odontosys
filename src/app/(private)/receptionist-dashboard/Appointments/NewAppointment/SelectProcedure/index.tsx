import styles from "./styles.module.scss"
import Label from "@/components/ui/Label"
import Button from "@/components/ui/Button"
import { ArrowLeft, ArrowRight, OctagonX } from "lucide-react"
import Spinner from "@/components/ui/Spinner"
import FeedbackMessage from "@/components/ui/FeedbackMessage"
import useSWR from "swr"
import fetcher from "@/services/fetcher"
import { useAppointmentContext } from "@/contexts/AppointmentContext"

interface IProcedure {
    id: number | null
    procedure: string | null
    durationMinutes: number | null
}

export default function SelectProcedure() {
    // Busca os procedimentos disponíveis
    const { data: procedures, error, isLoading } = useSWR("/api/procedure", fetcher, { revalidateOnFocus: false })
    const { state, dispatch } = useAppointmentContext()

    // Função para lidar com a mudança de seleção do procedimento
    function handleChange(e: React.ChangeEvent<HTMLSelectElement>) {
        const selectedName = e.target.value

        const selectedProcedure = procedures.find(
            (p: IProcedure) => p.procedure === selectedName
        )

        dispatch({
            type: "SET_PROCEDURE",
            payload: {
                procedure: selectedProcedure?.procedure ?? null,
                durationMinutes: selectedProcedure?.durationMinutes ?? null
            }
        })
    }

    // Função para voltar para a etapa anterior
    function handleBack() {
        dispatch({ type: "SET_STEP", payload: state.step - 1 })
    }

    // Função para avançar para a próxima etapa
    function handleNext() {
        dispatch({ type: "SET_STEP", payload: state.step + 1 })
    }

    return (
        <div className={styles.box}>
            {isLoading ? (
                <div><Spinner /></div>
            ) : error ? (
                <FeedbackMessage message={error} icon={<OctagonX />} />
            ) : (
                <>
                    <Label htmlFor="procedure-select" text="Escolha o Procedimento" />
                    <select id="procedure-select" onChange={handleChange} value={state.procedure || ""}>
                        <option disabled value="">Selecione um procedimento</option>
                        {procedures.map((item: IProcedure) => <option key={item.id} value={item.procedure!}>{item.procedure}</option>)}
                    </select>
                </>
            )}
            {/* Botões de navegação */}
            <div className={styles.boxBtns}>
                <Button text="Voltar" iconStart={<ArrowLeft />} onClick={handleBack} />
                <Button text="Próximo" iconEnd={<ArrowRight />} onClick={handleNext} disabled={!state.procedure} />
            </div>
        </div>
    )
}