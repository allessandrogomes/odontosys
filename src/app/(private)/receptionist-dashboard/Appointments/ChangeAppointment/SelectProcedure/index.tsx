import styles from "./styles.module.scss"
import Label from "@/components/ui/Label"
import Button from "@/components/ui/Button"
import { ArrowLeft, ArrowRight, OctagonX } from "lucide-react"
import Spinner from "@/components/ui/Spinner"
import FeedbackMessage from "@/components/ui/FeedbackMessage"
import useSWR from "swr"
import fetcher from "@/services/fetcher"
import { useChangeAppointmentContext } from "@/contexts/ChangeAppointmentContext"

interface IProcedure {
    id: number | null
    procedure: string | null
    durationMinutes: number | null
}

export default function SelectProcedure() {
    // Busca os procedimentos disponíveis
    const { data: procedures, error, isLoading } = useSWR<IProcedure[] | null>("/api/procedure", fetcher, { revalidateOnFocus: false })
    const { state, dispatch } = useChangeAppointmentContext()

    // Função para lidar com a mudança de seleção do procedimento
    function handleChange(e: React.ChangeEvent<HTMLSelectElement>) {
        const selectedName = e.target.value

        // Armazena em 'selectedProcedure' o procedimento selecionado pelo usuário
        if (procedures && procedures.length > 0) {
            const selectedProcedure = procedures.find(
                (p: IProcedure) => p.procedure === selectedName
            )

            // Caso o procedimento selecionado exista, atualiza o state do contexto alterando apenas o 'procedure' e 'durationMinutes'
            // Se 'selectedAppointment' não existir, retorna o valor atual da consulta (não altera)
            if (selectedProcedure?.procedure && selectedProcedure.durationMinutes) {
                dispatch({
                    type: "SET_SELECTED_APPOINTMENT",
                    payload: state.selectedAppointment
                        ? {
                            ...state.selectedAppointment,
                            procedure: selectedProcedure?.procedure,
                            durationMinutes: selectedProcedure?.durationMinutes
                        }
                        : null
                })
            }
        }
    }

    // Função para voltar para a etapa anterior
    function handleBack() {
        dispatch({ type: "SET_STEP", payload: 2 })
    }

    // Função para avançar para a próxima etapa
    function handleNext() {
        dispatch({ type: "SET_STEP", payload: 4 })
    }

    return (
        <div className={styles.box}>
            {isLoading ? (
                <div><Spinner /></div>
            ) : error ? (
                <FeedbackMessage message={error} icon={<OctagonX />} />
            ) : (
                <>
                    <Label htmlFor="procedure-select" text="Escolha o Procedimento" />
                    <select id="procedure-select" onChange={handleChange} value={state.selectedAppointment?.procedure || ""}>
                        <option disabled value="">Selecione um procedimento</option>
                        {procedures && procedures.map((item: IProcedure) => <option key={item.id} value={item.procedure!}>{item.procedure}</option>)}
                    </select>
                </>
            )}
            {/* Botões de navegação */}
            <div className={styles.boxBtns}>
                <Button text="Voltar" iconStart={<ArrowLeft />} onClick={handleBack} />
                <Button text="Próximo" iconEnd={<ArrowRight />} onClick={handleNext} disabled={!state.selectedAppointment?.procedure} />
            </div>
        </div>
    )
}