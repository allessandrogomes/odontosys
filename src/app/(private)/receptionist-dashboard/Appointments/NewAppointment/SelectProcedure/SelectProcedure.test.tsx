import { useAppointmentContext } from "@/contexts/NewAppointmentContext"
import useSWR from "swr"
import SelectProcedure from "."
import { fireEvent, render, screen } from "@testing-library/react"

// Mock do contexto
jest.mock("@/contexts/NewAppointmentContext", () => ({
    useAppointmentContext: jest.fn()
}))

// Mock do useSWR
jest.mock("swr")

describe("SelectProcedure", () => {
    const dispatchMock = jest.fn()
    const proceduresMock = [
        { id: 1, procedure: "Limpeza", durationMinutes: 30 },
        { id: 2, procedure: "Restauração", durationMinutes: 60 }
    ]

    beforeEach(() => {
        jest.clearAllMocks();
        (useAppointmentContext as jest.Mock).mockReturnValue({
            state: { procedure: "" },
            dispatch: dispatchMock
        });

        (useSWR as jest.Mock).mockReturnValue({
            data: proceduresMock,
            error: null,
            isLoading: false
        })
    })

    it("deve renderizar corretamente e mostrar os procedimentos no select (snapshot)", () => {
        const { container } = render(<SelectProcedure />)

        const select = screen.getByLabelText("Escolha o Procedimento") as HTMLSelectElement
        expect(select).toBeInTheDocument()
        expect(select.value).toBe("") // valor inicial
        expect(screen.getByText("Selecione um procedimento")).toBeInTheDocument()
        expect(screen.getByText("Limpeza")).toBeInTheDocument()
        expect(screen.getByText("Restauração")).toBeInTheDocument()

        expect(container).toMatchSnapshot()
    })

    it("deve disparar dispatch correto ao clicar em 'Voltar'", () => {
        render(<SelectProcedure />)
        const backButton = screen.getByText("Voltar")
        fireEvent.click(backButton)

        expect(dispatchMock).toHaveBeenCalledWith({ type: "SET_STEP", payload: 2 })
    })

    it("deve disparar dispatch correto ao selecionar um procedimento", () => {
        render(<SelectProcedure />)

        const select = screen.getByLabelText("Escolha o Procedimento") as HTMLSelectElement
        fireEvent.change(select, { target: { value: "Restauração" } })

        expect(dispatchMock).toHaveBeenCalledWith({
            type: "SET_PROCEDURE",
            payload: { procedure: "Restauração", durationMinutes: 60 }
        })
    })

    it("deve exibir o Spinner quando isLoading for true", () => {
        // Mock específico para este teste
        (useSWR as jest.Mock).mockReturnValueOnce({
            data: null,
            error: null,
            isLoading: true
        })

        render(<SelectProcedure />)
        const spinner = screen.getByTestId("spinner")
        expect(spinner).toBeInTheDocument()
    })

    it("deve exibir FeedbackMessage quando houver erro", () => {
        const errorMessage = "Erro ao buscar procedimentos"

        // Mock específico para este teste
        ;(useSWR as jest.Mock).mockReturnValue({
            data: null,
            error: errorMessage,
            isLoading: false
        })
        
        render(<SelectProcedure />)

        const feedback = screen.getByText(errorMessage)
        expect(feedback).toBeInTheDocument()
    })
})