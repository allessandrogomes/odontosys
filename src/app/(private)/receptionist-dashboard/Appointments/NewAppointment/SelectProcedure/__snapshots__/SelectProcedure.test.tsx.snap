// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SelectProcedure deve renderizar corretamente e mostrar os procedimentos no select (snapshot) 1`] = `
<div>
  <div
    class="box"
  >
    <label
      class="label"
      for="procedure-select"
    >
      Escolha o Procedimento
    </label>
    <select
      id="procedure-select"
    >
      <option
        disabled=""
        value=""
      >
        Selecione um procedimento
      </option>
      <option
        value="Limpeza"
      >
        Limpeza
      </option>
      <option
        value="Restauração"
      >
        Restauração
      </option>
    </select>
    <div
      class="boxBtns"
    >
      <button
        class="button false "
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-arrow-left"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m12 19-7-7 7-7"
          />
          <path
            d="M19 12H5"
          />
        </svg>
        Voltar
      </button>
      <button
        class="button disabled "
        disabled=""
        type="button"
      >
        Próximo
        <svg
          aria-hidden="true"
          class="lucide lucide-arrow-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 12h14"
          />
          <path
            d="m12 5 7 7-7 7"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
